credentials/src/main/java/com/epam/aidial/core/credentials/service/ResourceCredentialsService.java

```
@@ -44,20 +45,45 @@ public void addResourceCredentials(CredentialsDescriptor credentialsDescriptor,
         }
 
-        byte[] body = JsonMapperUtil.convertToString(resourceCredentials).getBytes(StandardCharsets.UTF_8);
-        byte[] encryptedBody = encrypt(credentialsDescriptor, body);
+        byte[] encryptedBody = encrypt(credentialsDescriptor, resourceCredentials);
         resourceService.putResourceBytes(credentialsDescriptor.toResourceDescriptor(), encryptedBody, EtagHeader.ANY);
         log.debug("Resource credentials for resourceId={}, bucket={} stored successfully",
                 credentialsDescriptor.getResourceId(), credentialsDescriptor.getBucketName());
     }
 
+    public boolean copyResourceCredentials(CredentialsDescriptor from, CredentialsDescriptor to,
+                                           CredentialsLevel credentialsLevel, boolean overwrite) {
+        log.debug("Copying resource credentials for resourceId={} from bucket={} to bucket={}",
+                from.getResourceId(), from.getBucketName(), to.getBucketName());
+
+        ResourceCredentials resourceCredentials = getResourceCredentials(from);
+        if (resourceCredentials == null) {
+            log.debug("No resource credentials found for resourceId={} from bucket={}",
+                    from.getResourceId(), from.getBucketName());
+            return false;
+        }
+        if (!resourceCredentials.getCredentialsLevel().equals(credentialsLevel)) {
+            log.debug("Resource credentials found for resourceId={} from bucket={} has level={}, required level={}",
+                    from.getResourceId(), from.getBucketName(), resourceCredentials.getCredentialsLevel(), credentialsLevel);
+            return false;
+        }
+
+        resourceCredentials.setResourceId(to.getResourceId());
+
+        byte[] encryptedBody = encrypt(to, resourceCredentials);
+        EtagHeader etag = overwrite ? EtagHeader.ANY : EtagHeader.NEW_ONLY;
+        resourceService.putResourceBytes(to.toResourceDescriptor(), encryptedBody, etag);
```

If the resource exists the method throws http exception 412 instead of returning false.

I would recommend using copyResource

---

copyResource does not allow modification of resource data during copying, but we need to update the credentials' resourceId. So it would be simpler to use putResourceBytes